packages/teleport/src/Discover/Shared/JoinTokenContext.tsx

```diff
@@ -91,40 +108,42 @@ export function useJoinToken(): {
               'internal resource ID is required to discover the newly added resource, but none was provided'
             );
           }
-
-          return token;
-        })
-        .then(response => {
-          result.response = response;
-
-          tokenContext.setJoinToken(response);
+          cachedJoinTokenResult.response = token;
+          tokenContext.setJoinToken(token);
           tokenContext.startTimer();
         })
-        .catch(error => (result.error = error)),
+        .catch(error => {
+          cachedJoinTokenResult.error = error;
+        }),
     };
 
-    return result;
+    return cachedJoinTokenResult;
   }
 
   useEffect(() => {
-    return () => abortController.abort();
+    return () => {
+      abortController.abort();
+      // result will be stored in memory which can refer to
+      // previously used or expired join tokens.
+      clearCachedJoinTokenResult();
+    };
   }, []);
 
-  if (result) {
-    if (result.error) {
-      throw result.error;
+  if (cachedJoinTokenResult) {
+    if (cachedJoinTokenResult.error) {
+      throw cachedJoinTokenResult.error;
     }
 
-    if (result.response) {
+    if (cachedJoinTokenResult.response) {
       return {
-        joinToken: result.response,
+        joinToken: cachedJoinTokenResult.response,
         reloadJoinToken: run,
         timedOut: tokenContext.timedOut,
         timeout: tokenContext.timeout,
       };
     }
 
-    throw result.promise;
+    throw cachedJoinTokenResult.promise;
```

Comment:
Just a side note: Suspense for data fetching is still experimental even in the latest v18.2. Our current version is 16.8.4. Personally, I think we should use it only after this API is stabilized.
https://17.reactjs.org/docs/concurrent-mode-suspense.html

Reply:
It's already in the codebase as it's working fine for our simplified needs. If this was more complicated and was throwing bugs, I'd remove it - we'll be on React 18 soon, so I'd rather keep it in (it's already in the codebase) than refactor it out

Reply:
> Also we're already using Suspense here (albeit with a null fallback, but I'm not entirely sure why) - https://github.com/gravitational/webapps/blob/master/packages/teleport/src/Main/Main.tsx#L88-L90

I think it was only used to support react.lazy components which are fine since suspense for code splitting was added in 16.6 https://github.com/facebook/react/blob/main/CHANGELOG.md#1660-october-23-2018

Reply:
I mean my concern here is that this place could break even on a patch react update

Reply:
i have no strong opinions, but if we are going to the refactoring route, can we do it later? (since it's working fine atm). or when we upgrade react, we will be mindful of the suspense feature. our team isn't that big so i'm pretty positive one of us will remember it (:

Reply:
Sure